cli/azd/cmd/middleware/hooks.go

```diff
@@ -0,0 +1,159 @@
+	"github.com/azure/azure-dev/cli/azd/pkg/ext"
+	"github.com/azure/azure-dev/cli/azd/pkg/input"
+	"github.com/azure/azure-dev/cli/azd/pkg/project"
+)
+
+type HooksMiddleware struct {
+	env           *environment.Environment
+	projectConfig *project.ProjectConfig
+	commandRunner exec.CommandRunner
+	console       input.Console
+	options       *Options
+}
+
+// Creates a new instance of the Hooks middleware
+func NewHooksMiddleware(
+	env *environment.Environment,
+	projectConfig *project.ProjectConfig,
+	commandRunner exec.CommandRunner,
+	console input.Console,
+	options *Options,
+) Middleware {
+	return &HooksMiddleware{
+		env:           env,
+		projectConfig: projectConfig,
+		commandRunner: commandRunner,
+		console:       console,
+		options:       options,
+	}
+}
+
+// Runs the Hooks middleware
+func (m *HooksMiddleware) Run(ctx context.Context, next NextFn) (*actions.ActionResult, error) {
+	if err := m.registerServiceHooks(ctx); err != nil {
+		return nil, fmt.Errorf("failed registering service hooks, %w", err)
+	}
+
+	return m.registerCommandHooks(ctx, next)
+}
+
+// Register command level hooks for the executing cobra command & action
+// Invokes the middleware next function
+func (m *HooksMiddleware) registerCommandHooks(ctx context.Context, next NextFn) (*actions.ActionResult, error) {
+	if m.projectConfig.Hooks == nil || len(m.projectConfig.Hooks) == 0 {
+		log.Println("project does not contain any command hooks.")
+		return next(ctx)
+	}
+
+	hooksManager := ext.NewHooksManager(m.projectConfig.Path)
+	hooksRunner := ext.NewHooksRunner(
+		hooksManager,
+		m.commandRunner,
+		m.console,
+		m.projectConfig.Path,
+		m.projectConfig.Hooks,
+		m.env.Environ(),
+	)
+
+	var actionResult *actions.ActionResult
+
+	commandNames := []string{m.options.Name}
+	commandNames = append(commandNames, m.options.Aliases...)
+
+	err := hooksRunner.Invoke(ctx, commandNames, func() error {
+		result, err := next(ctx)
+		if err != nil {
+			return err
+		}
+
+		actionResult = result
+		return nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return actionResult, nil
+}
+
+// Registers event handlers for all services within the project configuration
+// Runs hooks for each matching event handler
+func (m *HooksMiddleware) registerServiceHooks(ctx context.Context) error {
+	for serviceName, service := range m.projectConfig.Services {
+		// If the service hasn't configured any hooks we can continue on.
+		if service.Hooks == nil || len(service.Hooks) == 0 {
+			log.Printf("service '%s' does not require any command hooks.\n", serviceName)
+			continue
+		}
+
+		serviceHooksManager := ext.NewHooksManager(service.Path())
+		serviceHooksRunner := ext.NewHooksRunner(
+			serviceHooksManager,
+			m.commandRunner,
+			m.console,
+			service.Path(),
+			service.Hooks,
+			m.env.Environ(),
+		)
+
+		for hookName, hookConfig := range service.Hooks {
+			hookType, eventName, err := inferHookName(hookName, hookConfig)
+			if err != nil {
+				return fmt.Errorf(
```

Comment:
This error wrapping actually seems inverse of conventional wrapping? I wonder if the more detailed error inside `inferHookName` should be "Hooks must start with 'pre' or 'post'", and the outer error should be "unable to infer type from hook name '%s' for service '%s'",

and would read:

unable to infer hook 'prefoo' for service 'web': hooks must start with 'pre' or 'post' and end in a valid service event name. Examples: restore, package, deploy